src/linker/pipeline.py

```diff
@@ -14,25 +16,29 @@ def __init__(self, config: Config):
         self.config = config
         self.steps = self._get_steps()
         self.implementations = self._get_implementations()
-        self.runner = None  # Assigned later from set_runner method
-
-    def set_runner(self, runner: Callable) -> None:
-        self.runner = runner
 
-    def run(self, results_dir: Path):
-        if not self.runner:
-            raise RuntimeError("Runner has not been set.")
+    def run(self, runner: Callable, results_dir: Path) -> None:
         for implementation in self.implementations:
-            self.runner(
+            implementation.run(
+                runner=runner,
                 container_engine=self.config.container_engine,
                 input_data=self.config.input_data,
                 results_dir=results_dir,
-                step_name=implementation.step.name,
-                implementation_name=implementation.name,
-                implementation_dir=implementation.directory,
-                container_full_stem=implementation.container_full_stem,
             )
 
+    def _validate(self, schema: "PipelineSchema") -> None:
+        for idx, implementation in enumerate(self.implementations):
+            # Check that all steps are accounted for and in the correct order
+            if implementation.step_name != schema.steps[idx].name:
+                return False
+            # Check that container exists
+            if (
```

Comment:
Move this to implementation validate function and also validate that implementation truly implements the step the user is claiming it implements